<?xml version="1.0"?>
<launch>
    <!-- ============================ SETTINGS ============================= -->

    <!-- Node frequency: the frequency [Hz] which rate.sleep() function will
    attempt to maintain in the main_loop. -->
    <arg 
    name="node_frequency" 
    default="1000"/>

    <arg 
    name="robot_name" 
    default="my_gen3"/>

    <!-- Tracking mode: "hold" or "toggle". If "hold" mode controller tracking
    button should be constantly held, in "toggle" mode - press once to track
    input, press again to stop tracking. -->
    <arg 
    name="tracking_mode" 
    default="toggle"/>

    <!-- Compensate orientation: if "true" - on tracking start or on mode switch
    the difference between current EE orientation and the controller input will
    be calculated. If "false", real controller input orientation will be sent to
    the robot (less safe, can cause jumps on tracking start). -->
    <arg 
    name="compensate_orientation" 
    default="true"/>

    <!-- Convenience compensation [degrees] is only used when compensate_orientation is -->
    <!-- set to "false". Allows to adjust controller input to a more comfortable -->
    <!-- orientation. -->
    <arg 
    name="convenience_compensation" 
    default="[0.0, 0.0, 0.0]"/>

    <!-- Maximum allowed change between the current input position / orientation
    and previous commanded position [meters] / orientation [degrees]. Protects
    against too big input change within one control loop cycle (~750Hz). The
    value should be selected experimentally given an input device and desired
    control sensitivity. -->
    <arg
    name="maximum_input_position_change"
    default="0.1"/>

    <arg
    name="maximum_input_orientation_change"
    default="8"/>

    <!-- If any of dependency or optional subfiles should not be launched (or
    are launched separately) set corresponding argument to "false". -->
    <arg 
    name="launch_positional_control" 
    default="true"/>

    <!-- If set to "true" enable rosbag recording of nodes' outputs. -->
    <arg 
    name="enable_rosbag_recording" 
    default="true"/>

    <!-- ======================== END OF SETTINGS ========================== -->


    <!-- Positional and gripper control (dependency): -->
    <group if="$(arg launch_positional_control)">
        <include 
        file="$(find kinova_positional_control)/launch/positional_control.launch">
            <arg 
            name="robot_name" 
            value="$(arg robot_name)"/>
        </include>
    </group>

    <!-- Teleoperation module: -->
    <group ns="/$(arg robot_name)">
        <node 
        pkg="kinova_positional_control" 
        type="teleoperation.py" 
        name="teleoperation" 
        output="screen">
            <param 
            name="node_frequency" 
            value="$(arg node_frequency)"/>
            <param 
            name="robot_name" 
            value="$(arg robot_name)"/>
            <param 
            name="tracking_mode" 
            value="$(arg tracking_mode)"/>
            <param 
            name="compensate_orientation" 
            value="$(arg compensate_orientation)"/>
            <param 
            name="convenience_compensation" 
            value="$(arg convenience_compensation)"/>
            <param
            name="maximum_input_position_change"
            value="$(arg maximum_input_position_change)"/>
            <param
            name="maximum_input_orientation_change"
            value="$(arg maximum_input_orientation_change)"/>
        </node>   
    </group>

    <!-- Rosbag: -->
    <group if="$(arg enable_rosbag_recording)">
        <node 
        pkg="rosbag" 
        type="record" 
        name="rosbag"
        ns="/$(arg robot_name)/teleoperation"
        output="screen"
        args="
        /$(arg robot_name)/teleoperation/is_initialized
        /$(arg robot_name)/positional_control/input_pose
        -o $(find kinova_positional_control)/bagfiles/teleoperation/$(arg robot_name)
        ">
        </node>    
    </group>
</launch>